lib/semantics/check-omp-structure.cc

```diff
@@ -82,10 +103,55 @@ void OmpStructureChecker::Enter(const parser::OpenMPLoopConstruct &x) {
   // push a context even in the error case
   OmpContext ct{dir.source};
   ompContext_.push_back(ct);
+  checkState_.justSawLoopConstruct = true;
+}
+
+void OmpStructureChecker::Enter(const parser::ExecutableConstruct &x) {
+  if (checkState_.justSawLoopConstruct) {
+    if (const auto *doDir{
+            std::get_if<common::Indirection<parser::DoConstruct>>(&x.u)}) {
+      // save associated DoConstruct pointer
+      SetContextDo(doDir->value());
+      checkState_.inOmpLoopRegion = true;
+    } else {
+      context_.Say(GetContext().directiveSource,
+          "do-loop is expected after the %s directive"_err_en_US,
+          ContextDirectiveAsFortran());
+      ompContext_.pop_back();
+    }
+    checkState_.justSawLoopConstruct = false;
+  }
+
+  const auto *endDoDir{
+      std::get_if<common::Indirection<parser::OpenMPEndLoopDirective>>(&x.u)};
+  if (!endDoDir && checkState_.allowOptionalEndDirective) {
+    // END directive is not present, loop construct is complete
+    checkState_.allowOptionalEndDirective = false;
+  }
 }
 
-void OmpStructureChecker::Leave(const parser::OpenMPLoopConstruct &) {
-  ompContext_.pop_back();
+void OmpStructureChecker::Leave(const parser::ExecutableConstruct &x) {
+  if (const auto *doDir{
+          std::get_if<common::Indirection<parser::DoConstruct>>(&x.u)}) {
+    if (checkState_.inOmpLoopRegion) {
+      if (&doDir->value() == GetContext().doDirective) {
+        checkState_.inOmpLoopRegion = false;
+        checkState_.allowOptionalEndDirective = true;
+        // pop at the end of associated do-loop
+        ompContext_.pop_back();
+      }
+    }
+  }
+}
```

Comment:
You may want have a return after `ompContext_.pop_back();` and add a `checkState_.allowOptionalEndDirective=false` in the fallback path to avoid accepting the following program that have leave other constructs after the parallel do:

```
  N = 1024
  do i = 1, 10
    !$omp parallel do copyin(a)
    do j = 2*i*N, (2*i+1)*N
    enddo
  enddo
  !$omp end parallel do
end
```
Note that the outer `do` could be an `if` or whatever `ExecutableConstruct` and the `!$omp end parallel do` would also be currently accepted.

Reply:
This is a nice catch, I forget that the `Leave` of the outer `DoConstrcut` follows the `Leave` of the inner `DoConstruct`.